src/field_theory/minpoly.lean

```diff
@@ -386,38 +386,94 @@ by simpa [sub_eq_add_neg] using add_algebra_map hx (-a)
+  {s : S} (hs : is_integral R s)
+
+include hs
+
 /-- For GCD domains, the minimal polynomial over the ring is the same as the minimal polynomial
-over the fraction field. -/
-lemma gcd_domain_eq_field_fractions {A R : Type*} (K : Type*) [comm_ring A] [is_domain A]
-  [normalized_gcd_monoid A] [field K]
-  [comm_ring R] [is_domain R] [algebra A K] [is_fraction_ring A K]
-  [algebra K R] [algebra A R] [is_scalar_tower A K R] {x : R} (hx : is_integral A x) :
-  minpoly K x = (minpoly A x).map (algebra_map A K) :=
+over the fraction field. See `minpoly.gcd_domain_eq_field_fractions'` if `S` is already a
+`K`-algebra. -/
+lemma gcd_domain_eq_field_fractions :
+  minpoly K (algebra_map S L s) = (minpoly R s).map (algebra_map R K) :=
 begin
-  symmetry,
-  refine eq_of_irreducible_of_monic _ _ _,
+  refine (eq_of_irreducible_of_monic _ _ _).symm,
   { exact (polynomial.is_primitive.irreducible_iff_irreducible_map_fraction_map
-      (polynomial.monic.is_primitive (monic hx))).1 (irreducible hx) },
-  { have htower := is_scalar_tower.aeval_apply A K R x (minpoly A x),
-    rwa [aeval, eq_comm] at htower },
-  { exact (monic hx).map _ }
+      (polynomial.monic.is_primitive (monic hs))).1 (irreducible hs) },
+   { rw [aeval_map, aeval_def, is_scalar_tower.algebra_map_eq R S L, ← eval₂_map, eval₂_at_apply,
+      eval_map, ← aeval_def, aeval, map_zero] },
+  { exact (monic hs).map _ }
+end
+
+/-- For GCD domains, the minimal polynomial over the ring is the same as the minimal polynomial
+over the fraction field. Compared to `minpoly.gcd_domain_eq_field_fractions`, this version is useful
+if the element is in a ring that is already a `K`-algebra. -/
+lemma gcd_domain_eq_field_fractions' [algebra K S] [is_scalar_tower R K S] :
+  minpoly K s = (minpoly R s).map (algebra_map R K) :=
+begin
+  let L := fraction_ring S,
+  rw [← gcd_domain_eq_field_fractions K L hs],
+  refine minpoly.eq_of_algebra_map_eq (is_fraction_ring.injective S L)
+    (is_integral_of_is_scalar_tower _ hs) rfl
 end
 
 /-- For GCD domains, the minimal polynomial divides any primitive polynomial that has the integral
 element as root. -/
-lemma gcd_domain_dvd {A R : Type*} (K : Type*)
-  [comm_ring A] [is_domain A] [normalized_gcd_monoid A] [field K]
-  [comm_ring R] [is_domain R] [algebra A K]
-  [is_fraction_ring A K] [algebra K R] [algebra A R] [is_scalar_tower A K R]
-  {x : R} (hx : is_integral A x)
-  {P : A[X]} (hprim : is_primitive P) (hroot : polynomial.aeval x P = 0) :
-  minpoly A x ∣ P :=
+lemma gcd_domain_dvd (hinj : function.injective (algebra_map R S))
+  {P : R[X]} (hP : P ≠ 0) (hroot : polynomial.aeval s P = 0) : minpoly R s ∣ P :=
 begin
-  apply (is_primitive.dvd_iff_fraction_map_dvd_fraction_map K
-    (monic.is_primitive (monic hx)) hprim).2,
-  rw ← gcd_domain_eq_field_fractions K hx,
+  let K := fraction_ring R,
+  let L := fraction_ring S,
+  haveI : no_zero_smul_divisors R L,
```

Comment:
This deserves its own instance, in the form `no_zero_smul_divisors R S → no_zero_smul_divisors R (fraction_ring S)`.

Reply:
This already exists, it wasn't found because I used `function.injective` instead of `no_zero_smul_divisors`. It is now fixed.